### PR TITLE
[math] Add DARE overload that takes Q, R, and N cost matrices

### PR DESCRIPTION
**Pull request by calcmogul**
I've been using this overload to compute the LQR gains for implicit model following (using feedback to make one model behave like another, but without extra states hence "implicit").

The relations `Q_new = (A - A_ref)^T Q (A - A_ref)`, `R_new = B^T Q B + R`, and `N_new = (A - A_ref)^T QB` I used in the tests came from equation (C.9) in appendix C of https://file.tavsys.net/control/controls-engineering-in-frc.pdf. This should probably be documented in the test itself, but I'm not sure what your guidelines are for referencing external works, or if there's some other way you'd like me to do this.

I'm also open to a simpler way of defining the cost matrices, but this was the easiest way I was able to create a set of matrices that led to a positive semidefinite DARE solution; naively choosing N = 1 for some of the tests with the original A, B, Q, and R violated that invariant.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)

### CODE REVIEW

**Review by SeanCurtis-TRI — COMMENTED** [review]
+@hongkai-dai for feature review (feel free to pass it on to someone who you consider to be more suitable).

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MeAlTfJEpvluIzAj8_p:blyqo5x)* status: LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai)

**Review by hongkai-dai — COMMENTED** [review]
:lgtm: +@seancurtis-tri for Monday's platform review.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MeCujioB4Yg69Yw3YZX:bihqoph)* status: LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
-@seancurtis-tri +@ggould-tri (platform review switcheroo).

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MePpdQPCCfI6tWtwY1m:b-sbf15l)* status: LGTM missing from assignee ggould-tri(platform) (waiting on @ggould-tri)

**Review by ggould-tri — COMMENTED** [review]
:lgtm: with some stylistic notes below.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MePybHL3hgohsqLMX3-:b-et5qur)* status: 3 unresolved discussions (waiting on @calcmogul)

___
*[math/discrete_algebraic_riccati_equation.h, line 40 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePt8x6EIgTVGVYly5y:-MePt8x6EIgTVGVYly5z:b-t4s81h) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.h#L40)):*
> ```objc
> /// This is equivalent to:
> /// \f[
> /// A_{scr}'XA_{scr} - X - A_{scr}'XB(B'XB + R)^{-1}B'XA_{scr} + Q_{scr} = 0
> ```

This subscript "scr" notation isn't present elsewhere in Drake or in either of the referenced papers, so without an explanation here it conflicts with https://drake.mit.edu/styleguide/cppguide.html#General_Naming_Rules .

Can you provide an expansion of what "scr" means or stands for in the comments here?
___
*[math/discrete_algebraic_riccati_equation.cc, line 464 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePusLc7rv-fCGw7IjV:-MePusLc7rv-fCGw7IjW:b4smx7x) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.cc#L464)):*
> ```cpp
>     DRAKE_DEMAND(N.rows() == B.rows() && N.cols() == B.cols());
> 
>     Eigen::MatrixXd scrA = A - B * R.llt().solve(N.transpose());
> ```

minor:  Why does "scr" prefix A here but is a trailing subscript in the header comments?  It seems like `A_scr` would be the more natural phrasing.
___
*[math/test/discrete_algebraic_riccati_equation_test.cc, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePxjwT08JZ7yvYKF8M:-MePxjwT08JZ7yvYKF8N:b-u552la) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/test/discrete_algebraic_riccati_equation_test.cc#L53)):*
> ```cpp
>                (A.transpose() * X * B + N) * (B.transpose() * X * B + R).inverse() *
>                    (B.transpose() * X * A + N.transpose()) +
>                Q;
> ```

minor:  This math is fairly hard to follow.  At author's discretion (ie, dismiss this comment if you disagree), consider reformatting it to put one additive term per line and use leading rather than trailing operators, e.g.
```
  MatrixXd Y = (
      A.transpose() * X * A
      - X
      - (A.transpose() * X * B * (B.transpose() * X * B + R).inverse()
         * B.transpose() * X * A)
      + Q);
```

**Review by calcmogul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)* status: 3 unresolved discussions (waiting on @ggould-tri)

___
*[math/discrete_algebraic_riccati_equation.h, line 40 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePt8x6EIgTVGVYly5y:-MeQM9A92UP_tfBPqckk:b4gfyvc) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.h#L40)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

This subscript "scr" notation isn't present elsewhere in Drake or in either of the referenced papers, so without an explanation here it conflicts with https://drake.mit.edu/styleguide/cppguide.html#General_Naming_Rules .

Can you provide an expansion of what "scr" means or stands for in the comments here?
</blockquote></details>

The symbols I'd use instead in this case would be written as `\mathcal{A}` and `\mathcal{Q}` in LaTeX. It's basically a curly/stylistic A/Q to differentiate it from the A and Q already in use. That's what I used in section C.1 of the pdf, which is also consistent with the symbol Wikipedia uses on their LQR article. Looks like Doxygen's \f block is LaTeX, so I can just swap that in.
___
*[math/discrete_algebraic_riccati_equation.cc, line 464 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePusLc7rv-fCGw7IjV:-MeQMyGP4ESX2ILq4hBL:b-qagu5g) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.cc#L464)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

minor:  Why does "scr" prefix A here but is a trailing subscript in the header comments?  It seems like `A_scr` would be the more natural phrasing.
</blockquote></details>

The symbols I was going after were `\mathcal{A}`and `\mathcal{Q}`. I'm not sure what a good name in the code itself would be since it's not obvious what `mathcal` is either to people who don't know LaTeX.
___
*[math/test/discrete_algebraic_riccati_equation_test.cc, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePxjwT08JZ7yvYKF8M:-MeQNBQH3tKF-P8RHKhI:bykhr5x) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/test/discrete_algebraic_riccati_equation_test.cc#L53)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

minor:  This math is fairly hard to follow.  At author's discretion (ie, dismiss this comment if you disagree), consider reformatting it to put one additive term per line and use leading rather than trailing operators, e.g.
```
  MatrixXd Y = (
      A.transpose() * X * A
      - X
      - (A.transpose() * X * B * (B.transpose() * X * B + R).inverse()
         * B.transpose() * X * A)
      + Q);
```
</blockquote></details>

I agree. I'll reformat the original SolveDAREandVerify() as well for consistency.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @calcmogul)

___
*[math/discrete_algebraic_riccati_equation.h, line 40 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePt8x6EIgTVGVYly5y:-MeRW5jBCMEOQEZM-Sru:b-8xr7h9) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.h#L40)):*
<details><summary><i>Previously, calcmogul (Tyler Veness) wrote…</i></summary><blockquote>

The symbols I'd use instead in this case would be written as `\mathcal{A}` and `\mathcal{Q}` in LaTeX. It's basically a curly/stylistic A/Q to differentiate it from the A and Q already in use. That's what I used in section C.1 of the pdf, which is also consistent with the symbol Wikipedia uses on their LQR article. Looks like Doxygen's \f block is LaTeX, so I can just swap that in.
</blockquote></details>

Okay, nobody reading that PDF is going to know the LaTeX source of it -- certainly a non-LaTeX user isn't going to know -- but this at least renders well.
___
*[math/discrete_algebraic_riccati_equation.cc, line 464 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15352#-MePusLc7rv-fCGw7IjV:-MeRWmCyCXpJJJQ3XWVp:blzzrzz) ([raw file](https://github.com/robotlocomotion/drake/blob/7f2683bfea2410c43cfa9106c6b00b1647a4d23a/math/discrete_algebraic_riccati_equation.cc#L464)):*
<details><summary><i>Previously, calcmogul (Tyler Veness) wrote…</i></summary><blockquote>

The symbols I was going after were `\mathcal{A}`and `\mathcal{Q}`. I'm not sure what a good name in the code itself would be since it's not obvious what `mathcal` is either to people who don't know LaTeX.
</blockquote></details>

"_mathcal" is no more informative to a reader than "_scr" -- you're using LaTeX heiroglyphs and many/most Drake developers are not native TeX speakers.

A comment here like "This code follows the equations described at https://file.tavsys.net/control/controls-engineering-in-frc.pdf#page=237 .  The variables in that paper whose LaTeX symbols are script letters --  `\mathcal{A}` and `\mathcal{Q}`-- are represented here as `A_some_english_text` and `Q_some_english_text`" (where "some_english_text" is a human-readable description of what the scriptness actually means, which is not clear to me on my brief reading).

Ultimately we should write as though someone who isn't you and doesn't necessarily speak LaTeX is going to have to fix a bug in this code some day in the future; we should put ourselves in that person's shoes when naming our variables.

**Comment by calcmogul** [comment]
I pushed some naming and docs fixes (see the commit message for details). I went with a numeral subscript instead since that'll confuse fewer people. I don't think linking to the PDF will be necessary since this is just a simple change of variables to make the equation fit the form of the original DARE again.

I haven't been able to actually generate the docs yet, so I don't know if they render correctly. I don't run Ubuntu locally, so the bazel commands for Doxygen just bail with "unsupported OS". I tried copying the DARE source files into a fresh directory with a default Doxygen config, and doxygen processed the headers, but it didn't generate any documentation as if it ignored the comments. The Doxyfile template isn't usable as is because bazel does some find-replace to it.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @calcmogul)

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot ok to test

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @calcmogul)

**Comment by calcmogul** [comment]
OK, I just squashed the changes in that fix the build.

I had to make the DARE QRN overload use a Doxygen multiline comment instead of a series of single-line comments. The new DARE documentation's formula blocks use the bmatrix LaTeX environment, which requires backslashes to delimit the matrix rows. Drake uses single-line comments for Doxygen docs, so GCC was throwing a "multi-line comment" warning (turned into an error) when encountering the backslashes in the formula block. Using a Doxygen multiline comment avoids the problem.

This makes the Doxygen format inconsistent though, so an alternative solution would be passing `-Wno-multiline-comment` to the compiler. You'd think the compiler would be smart enough to not throw that warning if the next line was a comment anyway, but oh well.

**Review by ggould-tri — COMMENTED** [review]
@sherm1 as a dynamics expert do you have any advice on handling the doxygen/latex/gcc interaction here?  Or can you recommend someone on dynamics who knows these things?  I don't know how you are managing this interaction.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-Mee83fdBc-FHFxJBFDy:b577586)* status: :shipit: complete! all discussions resolved, LGTM from assignees ggould-tri(platform),hongkai-dai (waiting on @ggould-tri and @hongkai-dai)

**Review by sherm1 — COMMENTED** [review]
Yes. Our basic documentation philosophy is documented [here](https://drake.mit.edu/doxygen_cxx/group__multibody__notation.html). Most relevant in this case:
> We need developers to provide good local documentation for our mathematical algorithms in the form of Doxygen comments. That documentation will be best viewed after formatting by Doxygen, which employs other tools like LaTeX and Markdown for beautification. However, for effective development we need that documentation to be understandable and easily modified by developers who are working directly in the source code. Maximizing the post-Doxygen elegance of the formatted documentation can come at a steep cost in source code readability and writability. And difficult-to-write documentation often doesn't get written at all.

Making use of Unicode where possible rather than Latex is the easiest way to stay out of trouble. For example, here is a line formatted from the embedded Latex in this PR:


Here is that same line in Unicode: 
    AᵀXA − X − AᵀXB(BᵀXB+R)⁻¹BᵀXA + Q = 0

There is no difference in information content, yet this line is readable directly in the header file as well as in doxygen.

So my first response would be that most if not all of the Latex here is unnecessary and produces substandard documentation in the header file. It would be easy I think to provide documentation that would be readable in both places, with no loss of information content, and with no troubles with Doxygen and gcc.

For convenient generation of Unicode equivalents of Latex, use [unicode.net](https://www.unicodeit.net/).

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MeexZ2O86YwBhLPlaeR:by9nto5)* status: :shipit: complete! all discussions resolved, LGTM from assignees ggould-tri(platform),hongkai-dai (waiting on @ggould-tri and @hongkai-dai)

**Comment by calcmogul** [comment]
That's a really neat tool. I just cleaned up the one-liner equations. Unfortunately though, there isn't a unicode equivalent for matrix typesetting that'll render nice in Doxygen.

Some of the DARE implementation comments line up brackets manually, but Doxygen will mangle that unless we use a <pre> block or something. It won't render nice as Doxygen, but it's more readable at the source level.

**Comment by sherm1** [comment]
Yeah, sometimes a `<pre>` or `@verbatim` block is a reasonable fallback. You can still use Unicode in those.

**Comment by calcmogul** [comment]
OK, verbatim blocks are in now for the stuff that needs matrices or integration limit notation, and I switched the comment block back to `///`.

**Comment by sherm1** [comment]
BTW it is fine to use `/** */` comments in Drake -- we are quite flexible regarding comment format. The rule we follow though is that the comment format must be consistent within a single file.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 1 of 2 files at r4, 1 of 2 files at r5, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352)* status: :shipit: complete! all discussions resolved, LGTM from assignees ggould-tri(platform),hongkai-dai (waiting on @calcmogul)

**Review by calcmogul — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15352#-:-MefXKse7NbsFf7CQ1Bi:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees ggould-tri(platform),hongkai-dai (waiting on @calcmogul)

**Comment by jwnimmer-tri** [comment]
Thank you and good work @calcmogul !
